src/Features/LanguageServer/Protocol/Handler/Diagnostics/DocumentPullDiagnosticHandler.cs

```diff
@@ -39,10 +53,18 @@ protected override ImmutableArray<Document> GetOrderedDocuments(RequestContext c
             return context.Document == null ? ImmutableArray<Document>.Empty : ImmutableArray.Create(context.Document);
         }
 
-        protected override DiagnosticParams[]? GetPreviousResults(DocumentDiagnosticsParams diagnosticsParams)
-            => new[] { diagnosticsParams };
+        protected override Task<ImmutableArray<DiagnosticData>> GetDiagnosticsAsync(
+            Document document, Option2<DiagnosticMode> diagnosticMode, CancellationToken cancellationToken)
+        {
+            // We only support doc diagnostics for open files.
+            if (!document.IsOpen())
+                return SpecializedTasks.EmptyImmutableArray<DiagnosticData>();
 
-        protected override IProgress<DiagnosticReport[]>? GetProgress(DocumentDiagnosticsParams diagnosticsParams)
-            => diagnosticsParams.PartialResultToken;
+            // For open documents, directly use the IDiagnosticAnalyzerService.  This will use the actual snapshots
+            // we're passing in.  If information is already cached for that snapshot, it will be returned.  Otherwise,
+            // it will be computed on demand.  Because it is always accurate as per this snapshot, all spans are correct
+            // and do not need to be adjusted.
+            return _analyzerService.GetDiagnosticsAsync(document.Project.Solution, documentId: document.Id, cancellationToken: cancellationToken);
```

Comment:
this will re-calculate every time until we fix the fork on every request behavior right?

Reply:
yes.